### Keras a High Level API for Tensorflow 2.0

#### Using The Sequential Model

##### Loading  The Dataset

In [3]:
import tensorflow as tf
print("tensorflow Version is : {}".format(tf.__version__))

tensorflow Version is : 2.0.0


In [5]:
mnist_data = tf.keras.datasets.mnist
(train_x, train_y),(test_x, test_y) = mnist_data.load_data()

11493376/11490434 [==============================] - 17s 1us/step


#### Defining The Variables 

In [6]:
epochs = 10
batch_size = 32

In [7]:
# Normalize all the data points and cast the labels to int64
train_x, test_x = tf.cast(train_x/255.0, tf.float32), tf.cast(test_x/255.0, tf.float32)
train_y, test_y = tf.cast(train_y, tf.int64), tf.cast(test_y, tf.int64)

In [8]:
train_x

<tf.Tensor: id=1, shape=(60000, 28, 28), dtype=float32, numpy=
array([[[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       ...,

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        

In [9]:
test_x

<tf.Tensor: id=3, shape=(10000, 28, 28), dtype=float32, numpy=
array([[[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       ...,

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        

In [10]:
train_y

<tf.Tensor: id=5, shape=(60000,), dtype=int64, numpy=array([5, 0, 4, ..., 5, 6, 8], dtype=int64)>

#### Building The Architecture 

In [11]:
mnistmodel1 = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

#### Compiling The Model

In [13]:
optimiser_1 = tf.keras.optimizers.Adam()
mnistmodel1.compile(optimiser=optimiser_1, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'] )    

#### Fitting The Model 

In [14]:
mnistmodel1.fit(train_x, train_y, batch_size=32, epochs=5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 18s 306us/sample - loss: 0.2208 - accuracy: 0.9352
Epoch 2/5
60000/60000 [==============================] - 17s 285us/sample - loss: 0.1092 - accuracy: 0.9694
Epoch 3/5
60000/60000 [==============================] - 19s 313us/sample - loss: 0.0840 - accuracy: 0.9766
Epoch 4/5
60000/60000 [==============================] - 18s 294us/sample - loss: 0.0735 - accuracy: 0.9796
Epoch 5/5
60000/60000 [==============================] - 19s 311us/sample - loss: 0.0632 - accuracy: 0.9832


In [15]:
mnistmodel1.fit(train_x, train_y, batch_size = 50, epochs = 2)

Train on 60000 samples
Epoch 1/2
60000/60000 [==============================] - 12s 194us/sample - loss: 0.0505 - accuracy: 0.9869
Epoch 2/2
60000/60000 [==============================] - 11s 187us/sample - loss: 0.0426 - accuracy: 0.9885


In [16]:
mnistmodel1.fit(train_x, train_y, batch_size=100, epochs = 5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 7s 120us/sample - loss: 0.0340 - accuracy: 0.9909
Epoch 2/5
60000/60000 [==============================] - 7s 112us/sample - loss: 0.0269 - accuracy: 0.9925
Epoch 3/5
60000/60000 [==============================] - 7s 124us/sample - loss: 0.0244 - accuracy: 0.9932
Epoch 4/5
60000/60000 [==============================] - 8s 133us/sample - loss: 0.0209 - accuracy: 0.9938
Epoch 5/5
60000/60000 [==============================] - 8s 132us/sample - loss: 0.0193 - accuracy: 0.9942


In [17]:
mnistmodel1.evaluate(test_x, test_y)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.07385185020834924, 0.9835]

This Represent a loss of 0.03 and an accuracy of 0.9835 on the test data.

An accuracy of 0.98 means that out of 100 test data points 98 were on average correctly identified by the model.

#### The Socond Way to create a sequential model.

#### The alternative to passing a list of layers to the sequential model's constructor is to use the add method as follows for the same architecture:

### Building The Architecture & Compiling

In [22]:
mnistmodel2 = tf.keras.models.Sequential();
mnistmodel2.add(tf.keras.layers.Flatten())
mnistmodel2.add(tf.keras.layers.Dense(512, activation = 'relu'))
mnistmodel2.add(tf.keras.layers.Dropout(0.2))
mnistmodel2.add(tf.keras.layers.Dense(10, activation=tf.nn.softmax))
mnistmodel2.compile(optimiser = tf.keras.optimizers.Adam(), loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])

#### Fitting The mnistmodel2

In [23]:
mnistmodel2.fit(train_x, train_y, batch_size = 50, epochs = 8)

Train on 60000 samples
Epoch 1/8
60000/60000 [==============================] - 12s 198us/sample - loss: 0.2305 - accuracy: 0.9324
Epoch 2/8
60000/60000 [==============================] - 12s 208us/sample - loss: 0.1077 - accuracy: 0.9683
Epoch 3/8
60000/60000 [==============================] - 15s 242us/sample - loss: 0.0811 - accuracy: 0.9764
Epoch 4/8
60000/60000 [==============================] - 14s 226us/sample - loss: 0.0644 - accuracy: 0.9814
Epoch 5/8
60000/60000 [==============================] - 13s 221us/sample - loss: 0.0547 - accuracy: 0.9841
Epoch 6/8
60000/60000 [==============================] - 12s 205us/sample - loss: 0.0483 - accuracy: 0.9866
Epoch 7/8
60000/60000 [==============================] - 13s 217us/sample - loss: 0.0400 - accuracy: 0.9887
Epoch 8/8
60000/60000 [==============================] - 13s 218us/sample - loss: 0.0366 - accuracy: 0.9893


#### Evaluate The Mnistmodel2 

In [24]:
mnistmodel2.evaluate(test_x, test_y)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.08836529553374894, 0.9804]

## Keras Functional API

In [25]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist
(train_x, train_y), (test_x, test_y) = mnist.load_data()
train_x, test_x = train_x/255.0, test_x/255.0
epochs = 10

#### Building The Architecture

In [28]:
inputs = tf.keras.Input(shape = (28,28)) # Returns a 'Placeholder' tensor
x = tf.keras.layers.Flatten()(inputs)
x = tf.keras.layers.Dense(512, activation = 'relu', name = 'd1')(x)
x = tf.keras.layers.Dropout(0.2)(x)
predictions = tf.keras.layers.Dense(10, activation=tf.nn.softmax, name = 'd2')(x)
mnistmodel3 = tf.keras.Model(inputs = inputs, outputs = predictions)

### Compile And Fit

In [29]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel3.compile(optimizer=optimiser, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
mnistmodel3.fit(train_x, train_y, batch_size=50, epochs=epochs)

Train on 60000 samples
Epoch 1/10
60000/60000 [==============================] - 9s 145us/sample - loss: 0.2359 - accuracy: 0.9316
Epoch 2/10
60000/60000 [==============================] - 9s 145us/sample - loss: 0.1001 - accuracy: 0.9705
Epoch 3/10
60000/60000 [==============================] - 9s 153us/sample - loss: 0.0728 - accuracy: 0.9776
Epoch 4/10
60000/60000 [==============================] - 9s 147us/sample - loss: 0.0545 - accuracy: 0.9826
Epoch 5/10
60000/60000 [==============================] - 8s 138us/sample - loss: 0.0432 - accuracy: 0.9863
Epoch 6/10
60000/60000 [==============================] - 9s 155us/sample - loss: 0.0338 - accuracy: 0.9891
Epoch 7/10
60000/60000 [==============================] - 9s 146us/sample - loss: 0.0295 - accuracy: 0.9897
Epoch 8/10
60000/60000 [==============================] - 9s 148us/sample - loss: 0.0256 - accuracy: 0.9916
Epoch 9/10
60000/60000 [==============================] - 9s 154us/sample - loss: 0.0225 - accuracy: 0.9926
Epoch

### Evaluate The Mnistmodel3

In [30]:
mnistmodel3.evaluate(test_x, test_y)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.060063668483676155, 0.984]

### Subclassing The Keras Model Class

In [31]:
import tensorflow as tf
print("Tensorflow  Version is : {}". format(tf.__version__))

Tensorflow  Version is : 2.0.0


### Building The Subclass Architecture 

In [32]:
class MNISTModel(tf.keras.Model):
    def __init__(self,num_classes=10):
        super(MNISTModel, self).__init__()
        # Define Your Layer Here.
        inputs = tf.keras.Input(shape = (28,28))# Return a 'placeholder' tensor
        self.x0 = tf.keras.layers.Flatten()
        self.x1 = tf.keras.layers.Dense(512, activation = 'relu', name = 'd1')
        self.x2 = tf.keras.layers.Dropout(0.2)
        self.predictions = tf.keras.layers.Dense(10, activation=tf.nn.softmax, name = 'd2')
    
    def call(self,inputs):
        # This is where to define your forward pass
        # Using the layers previously defined in __init__
        x = self.x0(inputs)
        x = self.x1(x)
        x = self.x2(x)
        return self.predictions(x)


In [33]:
mnistmodel4 = MNISTModel()

### Compile & Fit

In [34]:
batch_size = 50
steps_per_epoch = len(train_x.numpy())//batch_size
print(steps_per_epoch)
mnistmodel4.compile(optimizer=tf.keras.optimizers.Adam(), loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
mnistmodel4.fit(train_x, train_y, batch_size = batch_size, epochs=epochs)

AttributeError: 'numpy.ndarray' object has no attribute 'numpy'